sema/subyt/api.py

```diff
@@ -0,0 +1,379 @@
+
+
+# TODO make a pandas source in sources.py
+
+
+class GeneratorSettings:
+    """Embodies all the actual possible modifiers to the process"""
+
+    _scheme: Dict[str, Dict] = {
+        "ignorecase": {
+            "default": True,
+            "description": (
+                "Make all keys lowercase so "
+                "to ignore case in key references"
+            ),
+        },
+        "flatten": {
+            "default": True,
+            "description": (
+                "Flatten hierarchical strcutures"
+                " by making hierarchical key references."
+            ),
+        },
+        "iteration": {
+            "default": True,
+            "description": (
+                "Perform the iteration outside of "
+                "the template to avoid looping inside of it"
+            ),
+        },
+    }
+    _negation: str = "no-"
+
+    @staticmethod
+    def describe() -> str:
+        return "\n".join(
+            [
+                "%30s: %s" % (key, val["description"])
+                for (key, val) in GeneratorSettings._scheme.items()
+            ]
+        )
+
+    def __init__(self, modifiers: str = None):
+        self._values = {
+            key: val["default"]
+            for (key, val) in GeneratorSettings._scheme.items()
+        }
+        self.load_from_modifiers(modifiers)
+
+    def load_from_modifiers(self, modifiers: str):
+        """
+        Parses the -m --mode string into actual properties
+        (no-)ig(norecase),(no-)fl(atten),(no-)it(eration)
+        """
+        if modifiers is None:
+            return
+        # else
+        set_parts: Dict[str, bool] = {
+            key: False for (key, val) in self._values.items()
+        }
+
+        for part in modifiers.split(","):
+            val: bool = True
+            if part.startswith(GeneratorSettings._negation):
+                val = False
+                part = part[len(GeneratorSettings._negation) :]
+            found = [
+                key
+                for (key, val) in self._values.items()
+                if key.startswith(part)
+            ]
+            assert (
+                len(found) == 1
+            ), "ambiguous modifier string '%s' matches list: %s" % (
+                part,
+                found,
+            )
+            key = found[0]
+            assert not set_parts[key], (
+                "ambiguous modifier string "
+                f"{part} matches key {key} which is already set"
+            )
+            set_parts[key] = True
+            self._values[key] = val
+
+    def as_modifier_str(self) -> str:
+        """
+        Reproduces the modifier string that declares these settings
+        """
+        return ",".join(
+            [
+                "no-" + key if not val else key
+                for (key, val) in self._values.items()
+            ]
+        )
+
+    def __repr__(self) -> str:
+        return "GeneratorSettings('%s')" % self.as_modifier_str()
```

Comment:
f-string?

best to search for ` % ` throughout? 
good chance to streamline our code-style